sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/EncoderUtils.scala

```diff
@@ -26,6 +27,29 @@ import org.apache.spark.sql.catalyst.util.{ArrayData, MapData}
 import org.apache.spark.sql.types.{ArrayType, BinaryType, BooleanType, ByteType, CalendarIntervalType, DataType, DateType, DayTimeIntervalType, Decimal, DecimalType, DoubleType, FloatType, IntegerType, LongType, MapType, ObjectType, ShortType, StringType, StructType, TimestampNTZType, TimestampType, UserDefinedType, VariantType, YearMonthIntervalType}
 import org.apache.spark.unsafe.types.{CalendarInterval, UTF8String, VariantVal}
 
+/**
+ * :: DeveloperApi ::
+ * Extensible [[AgnosticEncoder]] providing conversion extension points over type T
+ * @tparam T over T
+ */
+@DeveloperApi
+trait AgnosticExpressionPathEncoder[T]
```

Comment:
@chris-twiner can you give me an example of what you exactly are missing from the agnostic encoder framework. I'd rather solve this problem at that level than create an escape hatch to raw catalyst expressions. I am not saying that we should not do this, but I'd like to have a (small) discussion first.

My rationale for pushing for agnostic encoders is that I want to create a situation where the Classic and Connect Spark SQL interfaces are on par. Catalyst bespoke encoders - sort of - defeat that.

Reply:
@hvanhovell thanks for getting back to me, per the JIRA this is existing pre 4 functionality that is no longer fully working. 
 [Frameless](https://github.com/typelevel/frameless) for example uses an extensible encoder derivation with injection/ADT support to provide type safe usage at compile time.  [Quality](https://sparkutils.github.io/quality/latest/model/storage/) for example uses injections to store a result ADT efficiently, [this SO](https://stackoverflow.com/q/78331071/1028537) has a similar often occurring example that can be solved.  Lastly as the inbuilt encoders are not extensible you can bump into issues of it's derivation limitation ([java.util.Calendar for example](https://typelevel.org/frameless/TypedEncoder.html)).

wrt to fully a unified interface impl, that's understood but this change is a minimal requirement to re-enable frameless style usage.  I don't have any direct way to provide parity for connect yet (although your unification work provides a clear basis), I track it under [frameless #701](https://github.com/typelevel/frameless/issues/701), although to go further down that route I'd also need custom expressions support in connect (but that's off topic).

Reply:
Yeah, that makes sense. However, I do want to call out that this is mostly internal API; we do not guarantee any compatibility between (minor) releases. For that reason, historically, most spark libraries have to create per spark version releases. The issue here IMO falls in that category.

I understand that this is a somewhat frustrating and impractical stance. I am open to having this interface for now, provided that in the future we can migrate towards AgnosticEncoders. The latter probably requires us to add additional encoders to the agnostic framework (e.g. an encoder for union types...).

Reply:
wrt internal - very much understood, it's the price paid for the functionality and performance gains, as I target Databricks as well there is yet more fun - hence [shim](https://github.com/sparkutils/shim)'s complicated version support

Reply:
The formal stance here is that all of this is internal API. We are allowed to break this between releases. Libraries that use internal API typically need to create a version per Spark version because of this. The only solace is that AgnosticEncoders have much stronger backwards compatibility guarantees because of Connects backwards compatibility requirements.

I understand that this is frustrating and impractical. Because of this I am fine with allowing this backdoor for now, provided that we move to AgnosticEncoders for Spark 4.1. I am happy to help here. It would help to understand what is currently missing here. I also curious as to why this would require custom expression support.

Reply:
If Databricks compatibility something you want, then Agnostic Encoders are your friend.